abstract_replication_strategy: add for_each_natural_endpoint_until

bhalevy opened the pull request:
Currently, effective_replication_map::do_get_ranges accepts
a functor that traverses the natural endpoints of each token
to decide whether a token range should be returned or not.

This is done by copying the natural endpoints vector for
each token.  However, other than special strategies like
everywhere and local, the functor can be called on the
precalculated inet_address_vector_replica_set in the
replication_map and there's no need to copy it for each call.

for_each_natural_endpoint_until passes a reference to the function
down to the abstract replication strategy to let it work either
on the precalculated inet_address_vector_replica_set or
on a ad-hoc vector prepared by the replication strategy.
The function returns stop_iteration::yes when a match or mismatch
are found, or stop_iteration::no while it has no definite result.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4185/

michoecho reviewed (approved):
LGTM

asias commented on locator/abstract_replication_strategy.cc:
```diff
@@ -68,6 +68,12 @@ inet_address_vector_replica_set abstract_replication_strategy::get_natural_endpo
     return res->second;
 }
 
+bool abstract_replication_strategy::in_natural_endpoints(const noncopyable_function<bool(const inet_address_vector_replica_set&)>& func, const token& search_token, const effective_replication_map& erm) const {
```
We need to document what this function really does. The name in_natural_endpoints looks as if we can checking if a node is part of the natrual endpoints set. 

How about something like this:

get_natural_endpoints_and_traverse()

or 

get_natural_endpoints_and_run()

which reflects the fact we are getting the natural endpoints first and then run a function to against the natural endpoints.

bhalevy replied:
The point here is that we don't get the natural endpoints like the current get_natural_endpoints that returns a copy of the natural endpoints vector, but rather we're passing a const reference to the endpoints to the given function.  So how about `invoke_on_natural_endpoints`?

denesb replied:
I raise `for_each_natural_endpoints()`.

bhalevy replied:
I also initially thought of "for each" but it would make sense only if the function would have been invoked for each endpoint separately.
But we need access to the vector, e.g. to match only the first entry, which is the primary replica.

asias replied:
invoke_on_natural_endpoints is better than my version.   

Regarding the return value, does this make more sense?

void invoke_on_natural_endpoints([&invoke_result]  (const inet_address_vector_replica_set&) {
}, ... )

The return state belongs to the lambda we passed to.

bhalevy replied:
Sure. That makes sense, and I also contemplated about this myself.
It's a little harder to use this way since we'll need to keep a variable to hold the result, but that's not terrible.
Ideally we could define it as a template that gets the return type from the functor, but then the implementation will need to be in the header file and it's not worth the extra complication as this isn't called on the hot data path.

asias replied:
OK. I vote for no template as well ;)

bhalevy replied:
BTW, the `bool` return type is related to why I called this function `in_natural_endpoints` since the function passed to it was supposed to search something in the natural endpoints and to return true iff the condition was found.
But I agree there could be other uses for it that may return something else, like one of the endpoints.

bhalevy replied:
Maybe the template isn't that bad.  I'll take a stab at it...
The caller's code would be much more elegant if invoke_on_... just returns the value the function param returns.

bhalevy replied:
```
'virtual' cannot be specified on member function templates (fix available)clang(virtual_member_function_template)
```
:-(

bhalevy commented:
In v2 ([a53b1cf](https://github.com/scylladb/scylladb/commit/a53b1cfa6f26695904af4b00ae4af8c46b609d26) ): reimplemented based on review feedback.
- for_each_natural_endpoint_until now traverses the natural endpoints, calling the functor param for each endpoint at a time.
- this saves the costly creation of a `inet_address_vector_replica_set` in everywhere_replication_strategy (done for each token in `effective_replication_map::do_get_ranges`)
- Both the functor and for_each_natural_endpoint_until return stop_iteration result, allowing the functor to stop scanning when it has a definite boolean result.
  - This is reflected in the signature of the functions passed to `do_add_ranges` that now return a std::optional<bool> that is translated to stop_iteration::yes once engaged or to stop_iteration::no otherwise.

bhalevy commented:
Makes we wonder if we could optimize for local/everywhere strategies better in the upper layer.
And, if we're doing the right thing WRT primary ranges with everywhere replication strategy. My gut feeling is that the natural endpoints for each token in everywhere strategy should be rotated, rather than be in static order, since we determine the primary owner as the first in the vector. That works for simple and network_topology andcit seems like to should be extended to everywhere strategy.

Cc @kbr-scylla @avikivity

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4219/

kbr-scylla commented:
> My gut feeling is that the natural endpoints for each token in everywhere strategy should be rotated, rather than be in static order, since we determine the primary owner as the first in the vector. 

You mean primary replica. I don't remember what difference the notion of primary/non-primary replica makes. It was something with materialized views I think? cc @nyh 

If the order matters for whatever reason then I agree, the primary replica should be the owner of the token.

bhalevy commented:
Cc @tgrabiec do you have any insight on whether we should consider rotating the natural endpoints for each token so that the primary replica would be the token owner?

avikivity commented on locator/abstract_replication_strategy.cc:
```diff
@@ -68,6 +68,17 @@ inet_address_vector_replica_set abstract_replication_strategy::get_natural_endpo
     return res->second;
 }
 
+stop_iteration abstract_replication_strategy::for_each_natural_endpoint_until(const noncopyable_function<stop_iteration(const inet_address&)>& func, const token& search_token, const effective_replication_map& erm) const {
```
It's nicer to put the function as the last parameter, so there are no extra arguments past the lambda.

avikivity commented on locator/abstract_replication_strategy.cc:
```diff
@@ -136,7 +147,7 @@ insert_token_range_to_sorted_container_while_unwrapping(
 
 dht::token_range_vector
-effective_replication_map::do_get_ranges(noncopyable_function<bool(inet_address_vector_replica_set)> should_add_range) const {
+effective_replication_map::do_get_ranges(noncopyable_function<std::optional<bool>(const inet_address&)> should_add_range) const {
```
std::optional<bool> is just too complicated, you can't understand what the different levels mean.

bhalevy replied:
okay, it can be simplified by returning stop_iteration and getting a `bool&` for the result as a parameter.

avikivity commented:
Is this in some hot path?

bhalevy commented:
> Is this in some hot path?

Yes, for example: storage_proxy::do_query -> query_singular -> get_read_executor -> get_live_sorted_endpoints -> sort_endpoints_by_proximity -> topo.sort_by_proximity -> std::sort(..., topology::compare_endpoints)

bhalevy commented:
In v3 ([ad97810](https://github.com/scylladb/scylladb/commit/ad978101ad44c469d5da80d8ce1f3d2eb893162b)):
- changed order of params to for_each_natural_endpoint_until so the function comes last.
- changed signature to function passed to do_get_ranges to `stop_iteration(bool&, const inet_address&)` instead of returning `std::optional<bool>`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4242/

asias commented:
Looks good.

bhalevy commented:
@scylladb/scylla-maint please merge

nyh commented:
> > My gut feeling is that the natural endpoints for each token in everywhere strategy should be rotated, rather than be in static order, since we determine the primary owner as the first in the vector.
> 
> You mean primary replica. I don't remember what difference the notion of primary/non-primary replica makes. It was something with materialized views I think? cc @nyh

All nodes should obviously agree about the replicas owning some token range, but should also agree about the **order** of these replicas (which is the ring order). It is this order which allows all nodes to agree which of the RF replicas is the "primary" replica as the first replica in that order (the concept of primary replica is used, for example, in repair), the "secondary" replica (this is used in Alternator TTL, for example), and as you noted, materialized views use the entire order to match up base and view replicas. So we need to keep this order as it always was. The first node in the list should remain the primary owner of this range - not *this* node.

nyh commented on locator/abstract_replication_strategy.cc:
```diff
@@ -136,7 +147,7 @@ insert_token_range_to_sorted_container_while_unwrapping(
 
 dht::token_range_vector
-effective_replication_map::do_get_ranges(noncopyable_function<bool(inet_address_vector_replica_set)> should_add_range) const {
+effective_replication_map::do_get_ranges(noncopyable_function<stop_iteration(bool&, const inet_address&)> should_add_range) const {
```
I think your rewrite of get_ranges() and do_get_ranges() using the new for_each_natural_enpoint_until made the implementation very hard to understand - e.g., you left the functor name "should_add_range" but it's no longer a bool ("should").

bhalevy replied:
I'll try to come up with a better name for the functor parameter.
How about `should_add_range_to_endpoint`?

nyh replied:
No, my problem is with the word "should_add". It suggests the functor should return a boolean of whether to add something - not a boolean of whether to stop the iteration.

bhalevy replied:
Then `consider_add_range_to_endpoint` maybe?

nyh commented on locator/abstract_replication_strategy.cc:
```diff
@@ -155,13 +170,11 @@ effective_replication_map::do_get_ranges(noncopyable_function<bool(inet_address_
-            }
+    return do_get_ranges([ep] (bool& add_range, const inet_address& e) {
+        if ((add_range = (e == ep))) {
+            return stop_iteration::yes;
```
I'm really confused now. It **looks** to me as this stops the iteration - and stops adding things to the vector - as soon as we find one range. Isn't this incorrect? What am I missing?

Do we have a unit test for this function? Or a higher-level test that we believe exercises this function well?

bhalevy replied:
Your understanding is correct.
Note that previously we'd also `return true` and stop the iteration on first match:
```
        for (auto a : eps) {
            if (a == ep) {
                return true;
            }
```

This is tested pretty much by any topology change test.

nyh replied:
I just realized that there are **two** iterations happening here: There is an iteration over token ranges, and for each token range there is an iteration over the different replicas of this token range. The "stop_iteration" we're seeing here stops the inner iteration, not the outer iteration.

This confused me completely in my review.

I'm not sure what can be done to make this clearer, though.

I'll review again with my new understanding.

nyh commented on locator/abstract_replication_strategy.cc:
```diff
@@ -187,8 +200,9 @@ abstract_replication_strategy::get_ranges(inet_address ep, token_metadata_ptr tm
-        return eps.size() > 0 && eps[0] == ep;
+    return do_get_ranges([ep] (bool& add_range, const inet_address& e) {
+        add_range = (e == ep);
+        return stop_iteration::yes;
```
Again I'm really confused what this means - it seems like an error but maybe I'm just mistunderstanding.... Why stop after one iteration, don't we need to look at all the ranges?

bhalevy replied:
Previously, we also checked only the first entry:
```
        return eps.size() > 0 && eps[0] == ep;
```

The left part of the expression above is covered by the fact this functor is now called, meaning the vector has at least one entry,
The right part is covered by determining `add_value = (e == ep)` once, and stopping the iteration.

bhalevy commented:
> > > My gut feeling is that the natural endpoints for each token in everywhere strategy should be rotated, rather than be in static order, since we determine the primary owner as the first in the vector.
> > 
> > 
> > You mean primary replica. I don't remember what difference the notion of primary/non-primary replica makes. It was something with materialized views I think? cc @nyh
> 
> All nodes should obviously agree about the replicas owning some token range, but should also agree about the **order** of these replicas (which is the ring order). It is this order which allows all nodes to agree which of the RF replicas is the "primary" replica as the first replica in that order (the concept of primary replica is used, for example, in repair), the "secondary" replica (this is used in Alternator TTL, for example), and as you noted, materialized views use the entire order to match up base and view replicas. So we need to keep this order as it always was. The first node in the list should remain the primary owner of this range - not _this_ node.

Hmm, currently `everywhere_replication_strategy` returns the nodes as they are returned from `token_metadata::get_all_endpoints`, which in turn, returns `std::unordered_set<inet_address> _normal_token_owners`.
This works on all nodes since the hash function that effectively determines the order of elements in the unordered_set is the same on all nodes.

The question still is whether this needs to be rotated per search token such that the first node in the list of natural endpoints for a particular endpoint would be the primary owner of that token, for load balancing purposes.
If everywhere_replication_strategy would have followed network_topology_replication_strategy node selection algorithm, it would use the ring order, starting from the search token, and keep adding endpoints in the ring order until all are in the returned result.

We can do that once for the everywhere_replication_strategy instance, since it typically kept in sync with the token_metadata via the effective_replication_map, and rotate the natural endpoints list based on the search token.
That said, this will require some tricks to avoid inflating the replication map footprint to |tokens| * |nodes|.

bhalevy commented on locator/abstract_replication_strategy.cc:
```diff
@@ -68,6 +68,17 @@ inet_address_vector_replica_set abstract_replication_strategy::get_natural_endpo
+    const token& key_token = erm.get_token_metadata_ptr()->first_token(search_token);
+    auto res = erm.get_replication_map().find(key_token);
+    for (const auto& ep : res->second) {
+        if (func(ep)) {
```
stop_iteration is a bool_class so it is convertible to `bool`.
(Also for constructing as `stop_iteration(bool)`).

But I don't mind making this change if you think it makes the code clearer.

nyh replied:
Yes, I think even if this compiles it's very strange to return "yes" because we believe "return true" will not be clear enough, and one line up we think that if() will be clear enough.

nyh reviewed (approved):
In my first review, I got completely confused by the fact that most "iterations" changed by this patch were actually two nested iterations (going over all token ranges, and for each token range go over all the nodes in it), and the functions now use the concept of "stop_iteration" without explaining - not in comments nor in naming - which of the iterations is being stopped.

On second read, I understand what you did, and it looks correct, so I approve.
I'm still worried that this code isn't clear, but maybe it's a matter of style and it's not like I have a great suggestion on how to make it clearer.

bhalevy commented:
> > Hmm, currently `everywhere_replication_strategy` returns the nodes as they are returned from `token_metadata::get_all_endpoints`, which in turn, returns `std::unordered_set<inet_address> _normal_token_owners`. This works on all nodes since the hash function that effectively determines the order of elements in the unordered_set is the same on all nodes.
> > The question still is whether this needs to be rotated per search token
> 
> I think Scylla is gauge invariant :-) (if there are any physicists in the crowd, maybe they'll appreciate the analogy).
> 
> What I mean is that it does not matter how we order the replicas for **different** tokens - if it's the same order for all tokens or different order for different tokens. All we need is that for one specific token (or rather, token range a.k.a vnode), the decision on the order is consistent (same order) on all nodes. But we can change this decision from token range to token range.
> 
> So Scylla should be correct if you decide to rotate, or not rotate, the order for different tokens. After saying that, you may have a good reason to indeed do this rotation:
> 
> > such that the first node in the list of natural endpoints for a particular endpoint would be the primary owner of that token, for load balancing purposes.
> 
> Right. I don't think this will matter a lot but it might matter in a few specific cases. For example if node A is always the primary owner (first in the list) for all token ranges, "repair -pr" will end up doing all the repair work of this table on one node.

yes.  repair uses `get_primary_ranges_within_dc` which the first match in the local dc, so for everywhere_replication_strategy the first node (for whatever "first" is) will always be the same for all ranges and the repair -pr work will be distributed unevenly.  That was one of my main concerns in https://github.com/scylladb/scylladb/pull/12737#issuecomment-1420360408.
For regular reads and writes from tables with everywhere replication it shouldn't matter as we use HWLB to select replicas for read and we write to all nodes anyhow since the data is replicated everywhere.

> If this table can get big (I don't know if this happens for a everywhererelication table), it's not a good idea.
system.distributed tables are typically small so the impact is quite limited.
Choosing everywhere replication for big user tables is possible, I think.  It might be useful for read-mostly tables that are rarely updated.

Cc @asias

bhalevy commented:
v4 ([df3e521](https://github.com/scylladb/scylladb/commit/df3e5219294822bd83b44f4e420f96288ee2aa38)):
- renamed `should_add_range` functor parameter in `do_get_ranges` to `consider_range_for_endpoint`
- added commentary in calls to do_get_ranges explaining what the functor is supposed to do and why it stops the iteration.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4261/
